File: src/api/project.ts
import { invoke, PageResult } from "@/api";

export interface Project {
  id: number;
  name: string;
  description: string;
  gitlab_url: string;
  gitlab_token: string;
  gitlab_project_id: number;
  gitlab_branch: string;
  state: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectDetail {
  id: number;
  name: string;
  description: string;
  gitlab_url: string;
  gitlab_token: string;
  gitlab_project_id: number;
  gitlab_branch: string;
  state: number;
  created_at: string;
  updated_at: string;
  iteration_count: number;
  pending_sql_count: number;
  has_gitlab_config: boolean;
}

export interface GitCommit {
  id: number;
  project_id: number;
  commit_id: string;
  message: string;
  author_name: string;
  author_email: string;
  committed_at: string;
  additions: number;
  deletions: number;
  branch: string;
}

export const projectApi = {
  list: (params?: { page?: number; size?: number; keyword?: string }) =>
    invoke<PageResult<ProjectDetail>>("list_projects", { req: params || {} }),
  listAll: () =>
    invoke<Project[]>("list_all_projects"),
  detail: (id: number) =>
    invoke<ProjectDetail>("get_project_detail", { id }),
  add: (data: Partial<Project>) =>
    invoke<number>("add_project", { req: data }),
  update: (data: Partial<Project> & { id: number }) =>
    invoke<void>("update_project", { req: data }),
  delete: (id: number) =>
    invoke<void>("delete_project", { id }),
  syncCommits: (id: number) =>
    invoke<number>("sync_commits", { id }),
  getCommits: (projectId: number) =>
    invoke<GitCommit[]>("get_commits", { projectId }),
  listBranches: (params: { project_id?: number; gitlab_url?: string; gitlab_token?: string; gitlab_project_id?: number }) =>
    invoke<string[]>("list_gitlab_branches", { req: params }),
};

File: src-tauri/src/models/project.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Project {
    pub id: i32,
    pub name: String,
    pub description: String,
    pub gitlab_url: String,
    pub gitlab_token: String,
    pub gitlab_project_id: i32,
    pub gitlab_branch: String,
    pub enabled: i32,
    pub state: i32,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectListReq {
    #[serde(default)]
    pub page: Option<i64>,
    #[serde(default)]
    pub size: Option<i64>,
    #[serde(default)]
    pub keyword: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectAddReq {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub gitlab_url: Option<String>,
    #[serde(default)]
    pub gitlab_token: Option<String>,
    #[serde(default)]
    pub gitlab_project_id: Option<i32>,
    #[serde(default)]
    pub gitlab_branch: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectUpdateReq {
    pub id: i32,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub gitlab_url: Option<String>,
    #[serde(default)]
    pub gitlab_token: Option<String>,
    #[serde(default)]
    pub gitlab_project_id: Option<i32>,
    #[serde(default)]
    pub gitlab_branch: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectDetailRsp {
    #[serde(flatten)]
    pub project: Project,
    pub iteration_count: i64,
    pub pending_sql_count: i64,
    pub has_gitlab_config: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectEnabledUpdateReq {
    pub id: i32,
    pub enabled: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GitLabBranchReq {
    #[serde(default)]
    pub project_id: Option<i32>,
    #[serde(default)]
    pub gitlab_url: Option<String>,
    #[serde(default)]
    pub gitlab_token: Option<String>,
    #[serde(default)]
    pub gitlab_project_id: Option<i32>,
}
